akka-http2-support/src/main/scala/akka/http/scaladsl/Http2.scala

```diff
@@ -103,33 +105,65 @@ final class Http2Ext(private val config: Config)(implicit val system: ActorSyste
     settings:    ServerSettings,
     parallelism: Int,
     log:         LoggingAdapter
-  ): HttpRequest ⇒ Future[HttpResponse] = { req ⇒
+  )(implicit mat: Materializer): HttpRequest ⇒ Future[HttpResponse] = { req ⇒
     req.header[Upgrade] match {
       case Some(upgrade) if upgrade.protocols.exists(_.name equalsIgnoreCase "h2c") ⇒
-        // TODO remove duplication?
-        val serverLayer: Flow[ByteString, ByteString, Future[Done]] = Flow.fromGraph(
-          Flow[HttpRequest]
-            .watchTermination()(Keep.right)
-            // FIXME: parallelism should maybe kept in track with SETTINGS_MAX_CONCURRENT_STREAMS so that we don't need
-            // to buffer requests that cannot be handled in parallel
-            .via(Http2Blueprint.handleWithStreamIdHeader(parallelism)(handler)(system.dispatcher))
-            .joinMat(Http2Blueprint.serverStack(settings, log))(Keep.left))
-
-        // TODO consume the HTTP2-Settings header
-        // TODO consume the request entity and make sure it is inserted as an HTTP2 request
-        // https://http2.github.io/http2-spec/#rfc.section.3.2
-
-        Future.successful(
-          HttpResponse(
-            StatusCodes.SwitchingProtocols,
-            immutable.Seq[HttpHeader](
-              ConnectionUpgradeHeader,
-              UpgradeHeader,
-              // TODO add attributes from the request
-              UpgradeToOtherProtocolHeader(serverLayer)
+
+        log.debug("Got h2c upgrade request from HTTP/1.1 to HTTP2")
+
+        // https://http2.github.io/http2-spec/#Http2SettingsHeader 3.2.1 HTTP2-Settings Header Field
+        val upgradeSettings = req.headers.collect {
+          case raw: RawHeader if raw.lowercaseName == Http2SettingsHeader.name ⇒
+            Http2SettingsHeader.parse(raw.value)
+        }
+
+        upgradeSettings match {
+          // Must be exactly one
+          case immutable.Seq(Success(settingsFromHeader)) ⇒
+            // TODO remove duplication?
+
+            val injectedRequest =
+              if (req.method != HttpMethods.OPTIONS)
+                // inject the actual upgrade request with a stream identifier of 1
+                // https://http2.github.io/http2-spec/#rfc.section.3.2
+                Source.single(req.addHeader(Http2StreamIdHeader(1)))
+              else {
+                // No 100% sure about this but from the spec:
+                // If concurrency of an initial request with subsequent requests is important, an OPTIONS request
+                // can be used to perform the upgrade to HTTP/2, at the cost of an additional round trip.
```

Comment:
I think the point here is that an `OPTIONS` request will be empty and not 'congest' the first request. I suspect "After the empty line that terminates the 101 response, the server can begin sending HTTP/2 frames. These frames MUST include a response to the request that initiated the upgrade. The HTTP/1.1 request that is sent prior to upgrade is assigned a stream identifier of 1" still holds, even for OPTIONS requests - so we can remove the special case here but perhaps see if we should accept OPTIONS requests in the generated route

Reply:
Ok, I'll remove the special handling then.